<?xml version="1.0" encoding="UTF-8"?>
<!-- Launch a Gazebo simulation which demonstrates a swarm of EZ-RASSORs collaborating to mine regolith
     Written by Daniel Silva. -->

<launch>
  <arg name="robot_count" default="2"/>
  <arg name="spawn_x_coords" default="3 6"/>
  <arg name="spawn_y_coords" default="0 0"/>
  <arg name="spawn_z_coords" default="default"/>
  <arg name="enable_cameras" default="default"/>
  <arg name="enable_real_odometry" default="default"/>
  <arg name="swarm_control" default="default"/>

  <!-- Launch a rover's communication system -->
  <group ns="ezrassor$(arg robot_count)">
    <include file="$(find ezrassor_launcher)/launch/configurable_communication.launch">
      <arg name="control_methods" value="autonomy"/>
      <arg name="enable_cameras"
            value="$(arg enable_cameras)"
            unless="$(eval enable_cameras == 'default')"/>
      <arg name="enable_real_odometry"
            value="$(arg enable_real_odometry)"
            unless="$(eval enable_real_odometry == 'default')"/>
      <arg name="swarm_control"
            value="$(arg swarm_control)"
            unless="$(eval swarm_control == 'default')"/>
    </include>

     <!-- Launch the rover's control nodes. -->
    <include file="$(find ezrassor_sim_control)/launch/model_controls.launch"/>

    <!-- Spawn the rover's model. -->
    <include file="$(find ezrassor_sim_description)/launch/model.launch">
      <arg name="model_id" value="$(arg robot_count)"/>
      <arg name="spawn_x_coord"
           value="$(eval str(spawn_x_coords).split()[robot_count - 1])"
           unless="$(eval spawn_x_coords == 'default')"/>
      <arg name="spawn_y_coord"
           value="$(eval str(spawn_y_coords).split()[robot_count - 1])"
           unless="$(eval spawn_y_coords == 'default')"/>
      <arg name="spawn_z_coord"
           value="$(eval str(spawn_z_coords).split()[robot_count - 1])"
           unless="$(eval spawn_z_coords == 'default')"/>
    </include>

    <!-- Launch the rover's waypoint action client. -->
     <node pkg="ezrassor_swarm_control"
        name="waypoint_client"
        type="waypoint_client"
        output="screen">
       <param name="robot_num" value="$(arg robot_count)"/>
    </node>
  </group>

  <!-- Recursively spawn remaining rovers -->
  <include file="$(find ezrassor_swarm_control)/launch/spawn_rovers.launch"
           unless="$(eval robot_count == 1)">
    <arg name="robot_count" value="$(eval robot_count - 1)"/>
    <arg name="spawn_x_coords" value="$(arg spawn_x_coords)"/>
    <arg name="spawn_y_coords" value="$(arg spawn_y_coords)"/>
    <arg name="spawn_z_coords"
            value="$(arg spawn_z_coords)"
            unless="$(eval spawn_z_coords == 'default')"/>
    <arg name="enable_cameras"
            value="$(arg enable_cameras)"
            unless="$(eval enable_cameras == 'default')"/>
    <arg name="enable_real_odometry"
            value="$(arg enable_real_odometry)"
            unless="$(eval enable_real_odometry == 'default')"/>
    <arg name="swarm_control"
            value="$(arg swarm_control)"
            unless="$(eval swarm_control == 'default')"/>
  </include>
</launch>
